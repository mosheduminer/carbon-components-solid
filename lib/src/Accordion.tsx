import { Component, For, mergeProps, JSX } from "solid-js";
import { Dynamic } from "solid-js/web";
import { usePrefix } from "./internal/usePrefix";

export type AccordionProps = {
  align?: "start" | "end";
  class?: string;
  disabled?: boolean;
  size?: "sm" | "md" | "lg" | "xl";
} & JSX.HTMLAttributes<HTMLLIElement>;

export const Accordion: Component<AccordionProps> = (props) => {
  props = mergeProps({ align: "end" }, props);
  const prefix = usePrefix();

  return (
    <ul
      class={`${prefix}--accordion`}
      classList={{
        [props.class as string]: !!props.class,
        [`${prefix}--accordion--${props.align}`]: props.align !== undefined,
        [`${prefix}--accordion--${props.size}`]: props.size !== undefined,
      }}
    >
      {props.disabled ? (
        <For each={props.children as Component[]}>
          {(child) => <Dynamic component={child} />}
        </For>
      ) : (
        props.children
      )}
    </ul>
  );
};
