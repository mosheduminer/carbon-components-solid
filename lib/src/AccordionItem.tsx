import {
  Component,
  createSignal,
  splitProps,
  JSX,
  mergeProps,
  createEffect,
} from "solid-js";
import { settings } from "carbon-components";
import { ChevronRight } from "./icons/16";
import { match } from "./internal/keyboard/match";
import keys from "./internal/keyboard/keys";

const { prefix } = settings;

export type AccordionItemProps = {
  handleAnimationEnd?: JSX.EventHandler<HTMLLIElement, AnimationEvent>;
  class?: string;
  disabled?: boolean;
  onHeadingClick?: (props: { event: MouseEvent; isOpen: boolean }) => any;
  open?: boolean;
  title?: JSX.Element;
} & JSX.HTMLAttributes<HTMLLIElement>;

export const AccordionItem: Component<AccordionItemProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLLIElement>;
  [props, rest] = splitProps(props, [
    "class",
    "children",
    "disabled",
    "onHeadingClick",
    "open",
    "title",
    "handleAnimationEnd",
  ]);
  props = mergeProps({ open: false }, props);
  const [isOpen, setIsOpen] = createSignal(props.open);
  const [prevIsOpen, setPrevIsOpen] = createSignal(props.open);
  const [animation, setAnimation] = createSignal("");

  createEffect(() => {
    if (props.open !== prevIsOpen()) {
      setAnimation(isOpen() ? "collapsing" : "expanding");
      setIsOpen(props.open);
      setPrevIsOpen(props.open);
    }
  });

  function onClick(event: MouseEvent) {
    const nextValue = !isOpen();
    setAnimation(isOpen() ? "collapsing" : "expanding");
    setIsOpen(nextValue);
    if (props.onHeadingClick) {
      props.onHeadingClick({ isOpen: nextValue, event });
    }
  }

  // If the AccordionItem is open, and the user hits the ESC key, then close it
  function onKeyDown(event: KeyboardEvent) {
    if (isOpen() && match(event, keys.Escape)) {
      setIsOpen(false);
    }
  }

  function handleAnimationEnd(
    event: AnimationEvent & {
      currentTarget: HTMLLIElement;
      target: Element;
    }
  ) {
    if (props.handleAnimationEnd) {
      props.handleAnimationEnd(event);
    }
    setAnimation("");
  }

  return (
    <li
      class={`${prefix}--accordion__item`}
      classList={{
        [`${prefix}--accordion__item--active`]: isOpen(),
        [`${prefix}--accordion__item--${animation()}`]: !!animation(),
        [`${prefix}--accordion__item--disabled`]: props.disabled,
        [props.class!]: !!props.class,
      }}
      {...rest}
      onAnimationEnd={handleAnimationEnd}
    >
      <button
        disabled={props.disabled}
        aria-expanded={isOpen()}
        class={`${prefix}--accordion__heading`}
        onClick={(e) => onClick(e)}
        onKeyDown={onKeyDown}
        type="button"
      >
        <ChevronRight class={`${prefix}--accordion__arrow`} />
        <div class={`${prefix}--accordion__title`}>{props.title}</div>
      </button>
      <div class={`${prefix}--accordion__content`}>{props.children}</div>
    </li>
  );
};
