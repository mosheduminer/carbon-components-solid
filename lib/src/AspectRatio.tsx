import { Component, mergeProps, JSX, splitProps } from "solid-js";
import settings from "carbon-components/es/globals/js/settings";
import { Dynamic } from "solid-js/web";

const { prefix } = settings;

export type AspectRatioProps = {
  as?: Component | string;
  class?: string;
  ratio?: "16x9" | "9x16" | "2x1" | "1x2" | "4x3" | "3x4" | "1x1";
} & JSX.HTMLAttributes<HTMLElement>;

export const AspectRatio: Component<AspectRatioProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLElement>;
  [props, rest] = splitProps(props, ["class", "children", "as", "ratio"]);
  props = mergeProps({ as: "div", ratio: "1x1" }, props);
  return (
    <Dynamic
      component={props.as!}
      classList={{ [props.class!]: !!props.class }}
      class={`${prefix}--aspect-ratio ${prefix}--aspect-ratio--${props.ratio}`}
      {...rest}
    >
      {props.children}
    </Dynamic>
  );
};
