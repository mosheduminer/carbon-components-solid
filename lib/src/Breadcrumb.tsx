import { Component, JSX, splitProps } from "solid-js";
import { settings } from "carbon-components";

const { prefix } = settings;

export type BreadcrumbProps = {
  'aria-label'?: string;
  class?: string;
  noTrailingSlash?: boolean;
  ref?: HTMLElement;
} & JSX.HTMLAttributes<HTMLElement>;

export const Breadcrumb: Component<BreadcrumbProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLElement>;
  [props, rest] = splitProps(props, ["class", "children", "aria-label", "noTrailingSlash", "ref"]);

  return (
    <nav
      class={`${prefix}--breadcrumb`}
      classList={{ [`${prefix}--breadcrumb--no-trailing-slash`]: props.noTrailingSlash }}
      {...rest}
      ref={props.ref}>
      <ol classList={{[props.class!]: !!props.class}}>{props.children}</ol>
    </nav>
  )
}
