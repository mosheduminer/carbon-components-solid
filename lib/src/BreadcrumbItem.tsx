import {
  Component,
  Switch,
  Match,
  children,
  JSX,
  createEffect,
  splitProps,
} from "solid-js";
import { Dynamic } from "solid-js/web";
import { settings } from "carbon-components";
import { OverflowMenu } from "./OverflowMenu";
import { Link } from "./Link";
import { OverflowMenuHorizontal } from "./icons/32";

const { prefix } = settings;

export type BreadcrumbItemProps = {
  "aria-current"?: string | boolean;
  class?: string;
  href?: string;
  isCurrentPage: boolean;
  ref?: HTMLLIElement;
} & JSX.HTMLAttributes<HTMLLIElement>;

export const BreadcrumbItem: Component<BreadcrumbItemProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLLIElement>;
  [props, rest] = splitProps(props, [
    "aria-current",
    "class",
    "href",
    "isCurrentPage",
    "ref",
  ]);

  const classList = () => ({
    [`${prefix}--breadcrumb-item`]: true,
    [`${prefix}--breadcrumb-item--current`]:
      props.isCurrentPage && props["aria-current"] !== "page",
    [props.class!]: !!props.class,
  });
  const chlds = children(() => props.children) as () => JSX.Element;

  return (
    <Switch
      fallback={() => {
        createEffect(() => {
          (chlds()! as HTMLElement).classList.add(`${prefix}--link`);
          if (props["aria-current"]) {
            (chlds()! as HTMLElement).setAttribute("aria-current", "true");
          } else {
            (chlds()! as HTMLElement).removeAttribute("aria-current");
          }
        });
        return (
          <li classList={classList()} ref={props.ref} {...rest}>
            {chlds()}
          </li>
        );
      }}
    >
      <Match when={typeof chlds() === "string" && props.href}>
        <li classList={classList()} ref={props.ref} {...rest}>
          <Link href={props.href!} aria-current={props["aria-current"]}>
            {chlds()}
          </Link>
        </li>
      </Match>
      <Match when={typeof chlds() === "function"}>
        <li classList={classList()} {...rest}>
          <Dynamic
            component={chlds() as typeof OverflowMenu}
            menuOptionsClass={`${prefix}--breadcrumb-menu-options`}
            menuOffset={{ top: 10, left: 59 }}
            renderIcon={(props: JSX.HTMLAttributes<SVGSVGElement>) => (
              <OverflowMenuHorizontal
                class={`${prefix}--overflow-menu__icon`}
                {...props}
              />
            )}
          />
        </li>
      </Match>
    </Switch>
  );
};
