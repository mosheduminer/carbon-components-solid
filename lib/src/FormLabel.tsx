import { Component, JSX, splitProps } from "solid-js";
import settings from "carbon-components/es/globals/js/settings";

const { prefix } = settings;

export type FormLabelProps = {
  id?: string;
  class?: string;
} & JSX.HTMLAttributes<HTMLLabelElement>;

export const FormLabel: Component<FormLabelProps> = (props) => {
  let rest: JSX.HTMLAttributes<HTMLLabelElement>;
  [props, rest] = splitProps(props, ["class", "children", "id"]);
  return (
    <label
      for={props.id}
      class={`${prefix}--label`}
      classList={{ [props.class!]: !!props.class }}
      {...rest}
    >
      {props.children}
    </label>
  );
};
