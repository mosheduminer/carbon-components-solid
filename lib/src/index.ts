export * from "./Accordion";
export * from "./AccordionItem";
export * from "./AspectRatio";
export * from "./Breadcrumb";
export * from "./BreadcrumbItem";
export * from "./Button";
export * from "./ButtonSet";
export * from "./Checkbox";
export * from "./CodeSnippet";
export * from "./ComboBox";
export * from "./ComposedModal";
export * from "./ContentSwitch";
export * from "./ContentSwitcher";
export * from "./ContextMenu";
export * from "./Copy";
export * from "./CopyButton";
export * from "./DangerButton";
export * from "./DatePicker";
export * from "./DatePickerInput";
export * from "./FileUploader";
export * from "./Pagination";
export * from "./PaginationNav";
export * from "./ProgressBar";
export * from "./SkeletonText";
export * from "./Select";
export * from "./SelectItemGroup";
export * from "./SelectItem";
export * from "./StructuredList";
export * from "./Tag";
export * from "./TextInput";
