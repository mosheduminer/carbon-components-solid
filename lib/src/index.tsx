export * from "./Accordion";
export * from "./AccordionItem";
export * from "./AspectRatio";
export * from "./Breadcrumb";
export * from "./BreadcrumbItem";
export * from "./Button";
export * from "./ButtonSet";
export * from "./Checkbox";
export * from "./CodeSnippet";
export * from "./ComboBox";
export * from "./ComposedModal";
export * from "./ContentSwitch";
export * from "./ContentSwitcher";
export * from "./ContextMenu";
export * from "./Copy";
export * from "./CopyButton";
export * from "./DangerButton";
export * from "./DatePicker";
export * from "./DatePickerInput";
export * from "./FileUploader";
export * as icons from "./icons";
export * from "./SkeletonText";
export * from "./TextInput"
