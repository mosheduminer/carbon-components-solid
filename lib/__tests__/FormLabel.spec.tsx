import { render } from "solid-testing-library";
import { FormLabel } from '../src/FormLabel';

describe('FormLabel', () => {
  it('should render', () => {
    const { container } = render(() => <FormLabel />);
    expect(container).toMatchSnapshot();
  });
});