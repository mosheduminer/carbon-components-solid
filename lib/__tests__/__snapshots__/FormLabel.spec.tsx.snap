// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLabel should render 1`] = `
<div>
  <label
    class="bx--label"
  />
</div>
`;
